homeassistant/components/rainbird/__init__.py

```diff
@@ -61,47 +70,101 @@
-    client = AsyncRainbirdClient(async_get_clientsession(hass), server, password)
-    controller = AsyncRainbirdController(client)
-    try:
-        await controller.get_serial_number()
-    except RainbirdApiException as exc:
-        _LOGGER.error("Unable to setup controller: %s", exc)
-        return False
 
-    rain_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_sensor_state)
-    delay_coordinator = RainbirdUpdateCoordinator(hass, controller.get_rain_delay)
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up the config entry for Rain Bird."""
+    hass.data.setdefault(DOMAIN, {})
 
-    for platform in PLATFORMS:
-        hass.async_create_task(
-            discovery.async_load_platform(
-                hass,
-                platform,
-                DOMAIN,
-                {
-                    RAINBIRD_CONTROLLER: controller,
-                    SENSOR_TYPE_RAINSENSOR: rain_coordinator,
-                    SENSOR_TYPE_RAINDELAY: delay_coordinator,
-                    **controller_config,
-                },
-                config,
-            )
+    controller = AsyncRainbirdController(
+        AsyncRainbirdClient(
+            async_get_clientsession(hass),
+            entry.data[CONF_HOST],
+            entry.data[CONF_PASSWORD],
         )
+    )
+
+    try:
+        async with async_timeout.timeout(TIMEOUT_SECONDS):
+            serial_number = await controller.get_serial_number()
+    except (RainbirdApiException, asyncio.TimeoutError) as err:
+        raise ConfigEntryNotReady(f"Error talking to controller: {str(err)}") from err
+
+    device_info = DeviceInfo(
+        default_name=MANUFACTURER,
+        identifiers={(DOMAIN, serial_number)},
+        manufacturer=MANUFACTURER,
+    )
+    rain_coordinator = RainbirdUpdateCoordinator(
+        hass, "Rain", controller.get_rain_sensor_state
+    )
+    delay_coordinator = RainbirdUpdateCoordinator(
+        hass, "Rain delay", controller.get_rain_delay
+    )
+
+    hass.data[DOMAIN][entry.entry_id] = {
+        SERIAL_NUMBER: serial_number,
+        DEVICE_INFO: device_info,
+        RAINBIRD_CONTROLLER: controller,
+        SENSOR_TYPE_RAINSENSOR: rain_coordinator,
+        SENSOR_TYPE_RAINDELAY: delay_coordinator,
+    }
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    async def set_rain_delay(service: ServiceCall) -> None:
+        await controller.set_rain_delay(service.data[ATTR_DURATION])
+
+    hass.services.async_register(
+        DOMAIN,
+        SERVICE_SET_RAIN_DELAY,
+        set_rain_delay,
+        schema=SERVICE_SCHEMA_RAIN_DELAY,
+    )
+
     return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        hass.data[DOMAIN].pop(entry.entry_id)
+
+    if unload_ok and not hass.data[DOMAIN]:
```

Comment:
We should check if this is the last not loaded config entry.

https://github.com/home-assistant/core/blob/bdcccd9d830d45d965d3da1a907fdd89d991eac8/homeassistant/components/simplisafe/__init__.py#L463-L472